firmware/led-spiout.c

```diff
@@ -192,64 +185,128 @@ void led_set_spi_frequency(uint8_t frequency) {
     }
 }
 
+/* Sets all leds off, without using the interrupt */
+static void _led_all_off_no_interrupt() {
+    /* make sure we disabled the interrupt ! */
+    DISABLE_LED_WRITES;
 
-void led_init() {
+#define WAIT_SPI_TRANSMIT()     do { while (!(SPSR & _BV(SPIF))) ; } while (0)
+    SPDR = 0;
+    for (uint8_t i = 0; i < LED_START_FRAME_BYTES - 1; ++i) {
+        WAIT_SPI_TRANSMIT();
+        SPDR = 0;
+    }
+    for (uint8_t i = 0; i < NUM_LEDS; ++i) {
+        WAIT_SPI_TRANSMIT();
+        SPDR = BRIGHTNESS_MASK | 0;
+        WAIT_SPI_TRANSMIT();
+        SPDR = 0;
+        WAIT_SPI_TRANSMIT();
+        SPDR = 0;
+        WAIT_SPI_TRANSMIT();
+        SPDR = 0;
+    }
+    for (uint8_t i = 0; i < LED_END_FRAME_BYTES; ++i) {
+        WAIT_SPI_TRANSMIT();
+        SPDR = 0;
+    }
+    WAIT_SPI_TRANSMIT();
+#undef WAIT_SPI_TRANSMIT
+}
 
-    // Make sure all our LEDs start off dark
-    led_set_all_off();
+void led_init() {
 
     /* Set MOSI, SCK, SS all to outputs */
     DDRB = _BV(5)|_BV(3)|_BV(2);
     PORTB &= ~(_BV(5)|_BV(3)|_BV(2));
 
-    led_set_spi_frequency(led_spi_frequency);
+    /* Set all leds off as fast as possible.
+     *
+     * We don't use the SPI transfer interrupt here so we can synchronously turn
+     * off all leds now, then synchronously change SPI frequency without
+     * interference.
+     */
+    led_set_spi_frequency(LED_SPI_FREQUENCY_AT_INIT);
+    _led_all_off_no_interrupt();
 
-    /* Start transmitting the first byte of the start frame */
-    led_phase = START_FRAME;
+    /* Set the default led SPI frequency */
+    if (led_spi_frequency != LED_SPI_FREQUENCY_DEFAULT)
+        led_set_spi_frequency(LED_SPI_FREQUENCY_DEFAULT);
+
+    /* Trigger a first transmission */
     leds_dirty = 1;
-    SPDR = 0x0;
-    index = 1;
-    subpixel = 0;
+    ENABLE_LED_WRITES;
+    // Launch the very first transfer so SPI_STC_vect is called after that
+    // (We could also manually set SPIF)
+    // (An additional 8 zero bits start frame should not matter)
+    // SPDR = 0x00; // ! BUT don't do that since _led_all_off_no_interrupt() does it
 }
 
+typedef enum {
+    START_FRAME,
+    DATA_0,
+    DATA_1,
+    DATA_2,
+    DATA_3,
+    END_FRAME
```

Comment:
I understand what you're doing here, but something about it feels not quite right, possible because you're mixing what kind of frame we're in with  which subpixel of each LED we're in. It might be fixable with simple naming. How many instructions does this save us per LED? (Or, if you haven't looked, I can check it out later.)

Reply:
Would renaming them to `DATA_SUBFRAME_0` etc... help ?
I looked a bit deeper in the asm, saw a couple more good optims (7eb5ccd). But I think just looking at the C code we can see it reads and writes nearly half as much variables as before, and with less branching, so even if each additional `case` could generate a branch, all in all, it seems better I think.